src/tree/decisiontree.ts

```diff
@@ -0,0 +1,296 @@
+      if (this.nodes[i].is_left) {
+        this.nodes[this.nodes[i].parent_id].left_child_id = i
+      } else {
+        this.nodes[this.nodes[i].parent_id].right_child_id = i
+      }
+    }
+  }
+  predictProba(samples: number[][]): number[][] {
+    if (!this.is_built) {
+      throw new Error(
+        'Decision tree must be built with BuildTree method before predictions can be made.'
+      )
+    }
+    let leaf_node_ids = this.GetLeafNodes(samples)
+    let class_probabilities = []
+    let current_class_probabilities = []
+
+    for (let n_sample = 0; n_sample < leaf_node_ids.length; n_sample++) {
+      current_class_probabilities = []
+      let current_node_id = leaf_node_ids[n_sample]
+      for (let n_class = 0; n_class < this.nodes[0].value.length; n_class++) {
+        current_class_probabilities.push(
+          this.nodes[current_node_id].value[n_class] /
+            this.nodes[current_node_id].n_samples
+        )
+      }
+      class_probabilities.push(current_class_probabilities)
+    }
+    return class_probabilities
+  }
+  predict(samples: number[][]): int[] {
+    if (!this.is_built) {
+      throw new Error(
+        'Decision tree must be built with BuildTree method before predictions can be made.'
+      )
+    }
+    let leaf_node_ids = this.GetLeafNodes(samples)
+    let class_predictions = []
+
+    for (let n_sample = 0; n_sample < leaf_node_ids.length; n_sample++) {
+      let current_node_id = leaf_node_ids[n_sample]
+      class_predictions.push(argMax(this.nodes[current_node_id].value))
+      this.nodes[current_node_id].value
+    }
+    return class_predictions
+  }
+}
+
+export class DecisionTreeClassifier extends ClassifierMixin {
+  splitter_?: Splitter
+  stack_: NodeRecord[] = []
+  min_samples_leaf: int
+  max_depth_: int
+  min_samples_split: int
+  min_impurity_split_: number
+  n_features_: int = 0
+  tree_: DecisionTree
+  criterion_: ImpurityMeasure
+  max_features_: int
+  max_features_method: 'log2_method' | 'sqrt_method' | 'all_method'
+  feature_data_: number[][] = []
+  label_data_: number[] = []
+
+  constructor({
+    criterion = 'gini',
+    max_depth = Number.POSITIVE_INFINITY,
+    min_samples_split = 2,
+    min_samples_leaf = 1,
+    max_features = -1,
+    max_features_method = 'all_method',
+    min_impurity_split = 0.0
+  } = {}) {
+    super()
+    this.criterion_ = criterion as any
+    this.max_depth_ = max_depth
+    this.min_samples_split = min_samples_split
+    this.min_samples_leaf = min_samples_leaf
+    this.max_features_ = max_features
+    this.max_features_method = max_features_method as any
+    this.min_impurity_split_ = min_impurity_split
+    this.tree_ = new DecisionTree()
+  }
+
+  predict(feature_data: number[][]) {
+    return this.tree_.predict(feature_data)
+  }
+
+  predictProba(feature_data: number[][]) {
+    return this.tree_.predictProba(feature_data)
+  }
+  fit(feature_data: number[][], label_data: int[], samples_subset?: number[]) {
```

Comment:
Yup, it is a random forest thing. I think the reason they do it this way instead of the `X.gather` approach is that, you don't have to copy X ever. 

Like as soon as you read `X` into memory, you basically never touch it, and simply  pass around these `sample_subsets` which specify what rows you care about in both `X`, and `y`.